pkg/clients/service.go

```diff
@@ -0,0 +1,280 @@
+package clients
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+	hClient "github.com/ory/hydra-client-go/v2"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type PaginationLinksMeta struct {
+	First PaginationMeta `json:"first,omitempty"`
+	Last  PaginationMeta `json:"last,omitempty"`
+	Prev  PaginationMeta `json:"prev,omitempty"`
+	Next  PaginationMeta `json:"next,omitempty"`
+}
+
+type PaginationMeta struct {
+	Page string `json:"page,omitempty"`
+	Size int    `json:"size,omitempty"`
+}
+
+type ListClientsRequest struct {
+	PaginationMeta
+	Owner      string `json:"owner,omitempty"`
+	ClientName string `json:"client_name,omitempty"`
+}
+
+type ErrorOAuth2 struct {
+	Error            string `json:"error,omitempty"`
+	ErrorDescription string `json:"error_description,omitempty"`
+	StatusCode       int    `json:"-"`
+}
+
+type ServiceResponse struct {
+	Links        *PaginationLinksMeta
+	ServiceError *ErrorOAuth2
+	Resp         interface{}
+	Meta         map[string]string
+}
+
+type Service struct {
+	hydra HydraClientInterface
+
+	linksRegex *regexp.Regexp
+
+	tracer  trace.Tracer
+	monitor monitoring.MonitorInterface
+	logger  logging.LoggerInterface
+}
+
+func (s *Service) GetClient(ctx context.Context, clientID string) (*ServiceResponse, error) {
+	_, span := s.tracer.Start(ctx, "hydra.OAuth2Api.GetOAuth2Client")
+	defer span.End()
+
+	ret := NewServiceResponse()
+
+	c, resp, err := s.hydra.OAuth2Api().
+		GetOAuth2Client(ctx, clientID).
+		Execute()
+
+	if err != nil {
+		se, err := s.parseServiceError(resp)
+		if err != nil {
+			return nil, err
+		}
+		ret.ServiceError = se
+	}
+	ret.Resp = c
+	return ret, nil
+}
+
+func (s *Service) DeleteClient(ctx context.Context, clientID string) (*ServiceResponse, error) {
+	_, span := s.tracer.Start(ctx, "hydra.OAuth2Api.DeleteOAuth2Client")
```

Comment:
sorry i ve been missing a trick here, basically the context needs to be overridden (enhanced) and passed onwards so traces are propagated

```suggestion
	ctx, span := s.tracer.Start(ctx, "hydra.OAuth2Api.DeleteOAuth2Client")
```

this should be applied in every `tracer.Start`